File: 2022/july/07-07-2022/interleaving-string.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        @lru_cache(None)
        def isInterleaveHelper(i,j,k):
            if i==-1 and j==-1 and k==-1 :
                return True
            if (i == -1 and j == -1) and k>=0:
                return False
            match1 = False
            if i>=0 and k>=0 and s3[k] == s1[i]:
                match1 = isInterleaveHelper(i-1,j,k-1)
            match2 = False
            if j>=0 and k>=0 and s3[k] == s2[j]:
                match2 = isInterleaveHelper(i,j-1,k-1)
            return (match1 or match2)
        return isInterleaveHelper(len(s1)-1,len(s2)-1,len(s3)-1)

File: 2022/july/07-07-2022/interleaving-string.rs
impl Solution {
    pub fn is_interleave(s1: String, s2: String, s3: String) -> bool {
        let (l1, l2, l3) = (s1.len(), s2.len(), s3.len());
        if l1 + l2 != l3 {
            return false;
        }

        let (s1, s2, s3) = (s1.as_bytes(), s2.as_bytes(), s3.as_bytes());
        let mut dp: Vec<Vec<usize>> = vec![vec![0; l2 + 1]; l1 + 1];

        for i1 in 0..=l1 {
            for i2 in 0..=l2 {
                let i3 = i1 + i2;
                if i1 < l1 && s1[i1] == s3[i3] {
                    dp[i1 + 1][i2] = std::cmp::max(dp[i1 + 1][i2], dp[i1][i2] + 1);
                }
                if i2 < l2 && s2[i2] == s3[i3] {
                    dp[i1][i2 + 1] = std::cmp::max(dp[i1][i2 + 1], dp[i1][i2] + 1);
                }
            }
        }

        dp[l1][l2] == l3
    }
}